rust/ql/lib/codeql/rust/Concepts.qll

```diff
@@ -0,0 +1,136 @@
+/**
+ * Provides abstract classes representing generic concepts such as file system
+ * access or system command execution, for which individual framework libraries
+ * provide concrete subclasses.
+ */
+
+private import codeql.rust.dataflow.DataFlow
+private import codeql.threatmodels.ThreatModels
+
+/**
+ * A data flow source for a specific threat-model.
+ *
+ * Extend this class to refine existing API models. If you want to model new APIs,
+ * extend `ThreatModelSource::Range` instead.
+ */
+class ThreatModelSource extends DataFlow::Node instanceof ThreatModelSource::Range {
```

Comment:
Can't this simply be `final class ThreatModelSource = ThreatModelSource::Range`?

Reply:
I haven't worked with `Range` classes before.  My understanding is that in the frameworks we're expected to sometimes want to override both, so this can't be `final`, but .... I will admit I'm not really sure what the point of a separate class and `Range` is.

Reply:
I don't see why one would want to override both `ThreatModelSource` and `ThreatModelSource::Range`. I think we should change it like above.

Reply:
OK, I think I see what's going on here.  The intent of the "Range" pattern is that you can extend the abstract `Range` class to expand the class to cover new things (i.e. in this case to add new sources).  But if you want to define something in terms of `ThreatModelSource` (i.e. extend it the non-abstract way, as `ActiveThreatModelSource` does), you use the non-Range version.  Making it `final` actually _doesn't_ remove the latter possibility, thanks to "final extensions" (which I haven't knowingly used much before).

Change made.  The code is quite a lot cleaner as a result.  Other languages could implement the same change in their `Concepts.qll`.